Teach Python TS frontend to parse complex literals

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54542 Add support for real and imag tensor attributes
* #54541 [TS] Add complex support for more ops
* #54328 Allow creating SugaredValue for a complex valued IValue and deserialization logic for "infj" and "nanj" global constants
* #54089 Add JIT support for cmath unary ops
* **#52881 Teach Python TS frontend to parse complex literals**

**This PR adds:**
1. logic to parse complex constants (complex literals of the form `bj`)
2. logic to parse complex lists
3. support for complex constructors: `complex(tensor/int/float/bool, tensor/int/float/bool)`
4. Limited operator support
     - `add`, `sub`, `mul`, `torch.tensor`, `torch.as_tensor` 

**Follow-up work:**
1. Add complex support for unary and other registered ops.
2. support complex constructor with string as input (this is supported in Python eager mode).
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.
4. onnx doesn't support complex tensors, so we should error out with a clear and descriptive error message.

Differential Revision: [D27245059](https://our.internmc.facebook.com/intern/diff/D27245059)

anjali411 committed:
Teach Python TS frontend to parse complex literals

[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 commented on torch/csrc/jit/passes/erase_number_types.cpp:
```diff
@@ -39,8 +39,10 @@ static void EraseNumberTypesOnBlock(Block* block) {
       } break;
       case aten::Bool:
       case aten::Float:
+      case aten::Complex:
```
onnx only supports tensors, so this is used to turn all out number types into tensors

BowenBao replied:
ONNX does not support complex tensors yet. Could you leave out this change and have given onnx conversion properly error out with `aten::Complex` not supported?

anjali411 replied:
do you know if `createNumToTensor` was only added to support onnx?

BowenBao replied:
thanks! yea i think currently only onnx conversion is using that.

anjali411 commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1553,6 +1553,8 @@ inline TypePtr TensorType::fromNumberType(TypePtr typ) {
     return TensorType::createContiguous(at::kLong, at::kCPU, {});
   } else if (typ->isSubtypeOf(FloatType::get())) {
     return TensorType::createContiguous(at::kFloat, at::kCPU, {});
+  } else if (typ->isSubtypeOf(ComplexType::get())) {
```
for onnx:
used to create tensor from num

anjali411 commented on torch/csrc/jit/ir/node_hashing.cpp:
```diff
@@ -213,6 +213,10 @@ size_t HashNode::operator()(const Node* k) const {
         type->isSubtypeOf(NumberType::get()) &&
         k->kindOf(attr::value) == AttributeKind::f) {
       constant_hash = std::hash<double>{}(k->f(attr::value));
+    } else if (
```
updating `()` operator here helps access elements from the unordered_map `c_constants`

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"




[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


Follow-up work:
1. Add complex support for the remaining binary ops as well as unary ops, and other registered ops
2. Extensive TS support for complex lists and dicts
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

ezyang commented on test/jit/test_complex.py:
```diff
@@ -33,14 +35,34 @@ def test_pickle(self):
         self.assertEqual(loaded.b, [2 + 3j, 3 + 4j, -3j, -4])
         self.assertEqual(loaded.c, {2 + 3j : 2 - 3j, -4.3 - 2j: 3j})
+
+    @unittest.skipIf(IS_MACOS, "FIX ME: Tensors fail to compare equal on mac")
```
lol srsly?

malfet replied:
Oh, so this is still the issue?

ezyang commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -662,8 +667,13 @@ struct to_ir {
   ResolverPtr resolver;
-  std::unordered_map<int64_t, Value*> integral_constants;
-  std::unordered_map<double, Value*> fp_constants;
+  std::unordered_map<int64_t, Value*, std::hash<int>> integral_constants;
```
This looks like a semantic change. Previously, the hasher defaulted to `std::hash<int64_t>`, but you've chopped of some precision here. Was that intentional?

anjali411 replied:
good catch ... no it wasn't! I changed that back to `std::hash<int64_t>`

ezyang replied:
you need to do the same with double below

ezyang commented on torch/csrc/jit/frontend/lexer.h:
```diff
@@ -195,7 +195,12 @@ struct TORCH_API SharedParserData {
     torch::jit::strtod_c(startptr, &endptr);
-    *len = endptr - startptr;
+    // check if the number is complex valued
+    if (endptr != nullptr && *endptr == 'j') {
```
access is safe because string is assumed to be null terminated

ezyang commented:
This LGTM; letting TS team do signoff

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


Follow-up work:
1. Add complex support for the remaining binary ops as well as unary ops, and other registered ops
2. Extensive TS support for complex lists and dicts
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


Follow-up work:
1. Add complex support for the remaining binary ops as well as unary ops, and other registered ops
2. Extensive TS support for complex lists and dicts
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


Follow-up work:
1. Add complex support for the remaining binary ops as well as unary ops, and other registered ops
2. Extensive TS support for complex lists and dicts
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/runtime/register_ops_utils.h:
```diff
@@ -647,5 +647,142 @@ void listSetItem(Stack* stack);
+            pop(stack, a, b);                                                 \
+            push(stack, float_op);                                            \
+          },                                                                  \
+          aliasAnalysisFromSchema())
```
```suggestion
          aliasAnalysisFromSchema())

```

SplitInfinity reviewed (commented):
Nice! I left a few comments.

anjali411 commented on torch/csrc/jit/ir/constants.cpp:
```diff
@@ -15,8 +15,8 @@ bool insertableTensor(const at::Tensor& ten) {
   if (ivalue.isInt() || ivalue.isNone() || ivalue.isBool() ||
-      ivalue.isDouble() || ivalue.isString() || ivalue.isDevice() ||
-      ivalue.isEnum()) {
+      ivalue.isDouble() || ivalue.isComplexDouble() || ivalue.isString() ||
```
Hmm this is in line with the naming used for floating point types --> `isDouble`. I think the reason why we ultimately changed `ComplexDoubleType` to `ComplexType` was to be in line with the Python naming but `IValue` is a C++ construct not directly exposed in Python.

anjali411 commented on torch/csrc/jit/ir/attributes.h:
```diff
@@ -19,10 +19,26 @@ using ::c10::Symbol;
 
 constexpr int max_tensor_display_size = 10;
 
-enum class AttributeKind { f, fs, i, is, s, ss, t, ts, g, gs, ty, tys, ival };
```
>Does the addition of 2-3 extra characters really cause the formatting below?

I will add `cs` in a follow-up PR

>Also, why is there is no cs AttributeKind?

I will add it in a follow-up PR!

anjali411 commented on test/test_jit.py:
```diff
@@ -7378,10 +7379,11 @@ def func():
+                torch.as_tensor(input, dtype=torch.int32)
 
-        inp = torch.randn(3, 4)
+        inp = torch.randn(3, 4, dtype=torch.cfloat)
```
yeah it just uses the `copy` method which discards the imaginary values when copying complex values into real valued tensors.

anjali411 commented on torch/csrc/jit/frontend/tree_views.h:
```diff
@@ -867,12 +868,19 @@ struct Const : public Expr {
     tree_->matchNumSubtrees(TK_CONST, 1);
   }
   bool isFloatingPoint() const {
+    // Complex values consist of a real and an imaginary floating point value
```
yeah good point! updated the description to avoid any confusion

anjali411 replied:
actually removed the comment since I already have a detailed explanation in the comment below

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


Follow-up work:
1. Add complex support for the unary ops, and other registered ops
2. Extensive TS support for complex lists and dicts
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

anjali411 commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -336,6 +345,8 @@ RegisterOperators reg(
            pop(stack, scalar);
            if (scalar.isDouble()) {
              push(stack, std::move(scalar));
+           } else if (scalar.isComplexDouble()) {
+             push(stack, scalar.toComplexDouble().real());
```
yeah this is in line with the python behavior while casting complex numbers to floats

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


Follow-up work:
1. Add complex support for the unary ops, and other registered ops
2. Extensive TS support for complex lists and dicts
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

anjali411 commented on torch/csrc/jit/runtime/register_special_ops.cpp:
```diff
@@ -301,6 +301,13 @@ RegisterOperators reg({
                 bool,
                 bool,
                 at::empty({}, at::CPU(at::kBool).options()).fill_(scalar_val))
+                DEFINE_TORCH_TENSOR_OP(
+                    complex,
+                    c10::complex<double>,
+                    at::native::scalar_tensor(
+                        scalar_val,
+                        at::device(at::kCPU).dtype(
+                            c10::get_default_complex_dtype())))
```
yeah! I extended an existing test in `test_jit.py` to also run for complex types

anjali411 commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -1778,6 +1783,8 @@ Value* Graph::insertToList(Value* v, TypePtr type) {
     elem_ty = 1;
   } else if (ptr == BoolType::get()) {
     elem_ty = 2;
+  } else if (ptr == ComplexType::get()) {
+    elem_ty = 3;
```
done!

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


Follow-up work:
1. Add complex support for the unary ops, and other registered ops
2. Extensive TS support for complex lists and dicts
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


Follow-up work:
1. Add complex support for the unary ops, and other registered ops
2. Extensive TS support for complex lists and dicts
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


Follow-up work:
1. Add complex support for the unary ops, and other registered ops
2. Extensive TS support for complex lists and dicts
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


Follow-up work:
1. Add complex support for the unary ops, and other registered ops
2. Extensive TS support for complex lists and dicts
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


Follow-up work:
1. Add complex support for the unary ops, and other registered ops
2. Extensive TS support for complex lists and dicts
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


**This PR adds:**
1. logic to parse complex constants (complex literals of the form `bj`)
2. logic to parse complex lists
3. support for complex constructors: `complex(tensor/int/float/bool, tensor/int/float/bool)`
4. partial complex support for onnx (num -> tensor / tensor -> num)
5. Limited operator support
     - `add`, `sub`, `mul`, `torch.tensor`, `torch.as_tensor` 

**Follow-up work:**
1. Add complex support for unary and other registered ops.
2. support complex constructor with string as input (this is supported in Python eager mode).
2. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


**This PR adds:**
1. logic to parse complex constants (complex literals of the form `bj`)
2. logic to parse complex lists
3. support for complex constructors: `complex(tensor/int/float/bool, tensor/int/float/bool)`
4. partial complex support for onnx (num -> tensor / tensor -> num)
5. Limited operator support
     - `add`, `sub`, `mul`, `torch.tensor`, `torch.as_tensor` 

**Follow-up work:**
1. Add complex support for unary and other registered ops.
2. support complex constructor with string as input (this is supported in Python eager mode).
2. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

SplitInfinity commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1542,7 +1542,7 @@ def _make_tensor_helper(shape, low=None, high=None):
                                 dtypes=[torch.cfloat, torch.cdouble]),
                        SkipInfo('TestCommon', 'test_variant_consistency_jit',
-                                dtypes=[torch.cfloat, torch.cdouble, torch.bfloat16]),
+                                dtypes=[torch.bfloat16]),
```
❤️

SplitInfinity commented on torch/package/_mock.py:
```diff
@@ -1,23 +1,84 @@
-
-_magic_methods = ['__subclasscheck__', '__hex__', '__rmul__', '__complex__',
```
Can you keep the old formatting?

anjali411 replied:
there are the clang suggested formatted changes

SplitInfinity replied:
I would still undo the formatting changes. Also, this code isn't technically part of TorchScript, so you may not even need to modify it.

anjali411 replied:
removed!

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1236,6 +1236,31 @@ RegisterOperators reg2({
+        actual_type_a a;                                                 \
+        actual_type_b b;                                                 \
+        pop(stack, a, b);                                                \
+        auto comp = c10::complex<double>(a.item<double>(), b);           \
```
Do we need a check for whether the `Tensor` has only one element?

anjali411 replied:
I think item already does that check so it's not needed

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1236,6 +1236,31 @@ RegisterOperators reg2({
+            actual_type_b a;                                             \
+            actual_type_a b;                                             \
+            pop(stack, a, b);                                            \
+            auto comp = c10::complex<double>(a, b.item<double>());       \
```
Do we need a check for whether the Tensor has only one element?

anjali411 replied:
I think item already does that check so it's not needed

SplitInfinity commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -500,8 +500,6 @@ struct Environment {
           {"chr", std::make_shared<BuiltinFunction>(aten::chr, at::nullopt)},
           {"bin", std::make_shared<BuiltinFunction>(aten::bin, at::nullopt)},
           {"pow", std::make_shared<BuiltinFunction>(aten::pow, at::nullopt)},
-          {"complex",
-           std::make_shared<BuiltinFunction>(aten::complex, at::nullopt)},
```
Why did you get rid of this? To do in a followup diff?

anjali411 replied:
this was added in https://github.com/pytorch/pytorch/pull/53227 in an effort to support complex constructor for some inputs. However, with the change above, more input variants are supported

SplitInfinity reviewed (commented):
Nits

anjali411 commented on torch/csrc/jit/frontend/function_schema_parser.cpp:
```diff
@@ -173,6 +173,8 @@ struct SchemaParser {
         if ("float" == text) {
           return static_cast<int64_t>(at::kFloat);
+        } else if ("complex" == text) {
+          return static_cast<int64_t>(at::kComplexFloat);
```
just made the change based on what's being done for floating point

anjali411 commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -162,6 +162,10 @@ void Node::printAttrValue(std::ostream& out, const Symbol& name) const {
       printAttribute(out, c(name));
       break;
+    case AttributeKind::cs:
+      // TODO(@anjali411): fix this
+      AT_ASSERT(false);
```
some builds fail if 'cs' is not handled in switch case

SplitInfinity replied:
Ah

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


**This PR adds:**
1. logic to parse complex constants (complex literals of the form `bj`)
2. logic to parse complex lists
3. support for complex constructors: `complex(tensor/int/float/bool, tensor/int/float/bool)`
4. partial complex support for onnx (num -> tensor / tensor -> num)
5. Limited operator support
     - `add`, `sub`, `mul`, `torch.tensor`, `torch.as_tensor` 

**Follow-up work:**
1. Add complex support for unary and other registered ops.
2. support complex constructor with string as input (this is supported in Python eager mode).
2. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


**This PR adds:**
1. logic to parse complex constants (complex literals of the form `bj`)
2. logic to parse complex lists
3. support for complex constructors: `complex(tensor/int/float/bool, tensor/int/float/bool)`
4. Limited operator support
     - `add`, `sub`, `mul`, `torch.tensor`, `torch.as_tensor` 

**Follow-up work:**
1. Add complex support for unary and other registered ops.
2. support complex constructor with string as input (this is supported in Python eager mode).
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.
4. onnx doesn't support complex tensors, so we should error out with a clear and descriptive error message.

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


**This PR adds:**
1. logic to parse complex constants (complex literals of the form `bj`)
2. logic to parse complex lists
3. support for complex constructors: `complex(tensor/int/float/bool, tensor/int/float/bool)`
4. Limited operator support
     - `add`, `sub`, `mul`, `torch.tensor`, `torch.as_tensor` 

**Follow-up work:**
1. Add complex support for unary and other registered ops.
2. support complex constructor with string as input (this is supported in Python eager mode).
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.
4. onnx doesn't support complex tensors, so we should error out with a clear and descriptive error message.

[ghstack-poisoned]

anjali411 commented:
I'll add proper error checking for onnx in a follow-up PR!

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


**This PR adds:**
1. logic to parse complex constants (complex literals of the form `bj`)
2. logic to parse complex lists
3. support for complex constructors: `complex(tensor/int/float/bool, tensor/int/float/bool)`
4. Limited operator support
     - `add`, `sub`, `mul`, `torch.tensor`, `torch.as_tensor` 

**Follow-up work:**
1. Add complex support for unary and other registered ops.
2. support complex constructor with string as input (this is supported in Python eager mode).
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.
4. onnx doesn't support complex tensors, so we should error out with a clear and descriptive error message.

Differential Revision: [D27245059](https://our.internmc.facebook.com/intern/diff/D27245059)

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


**This PR adds:**
1. logic to parse complex constants (complex literals of the form `bj`)
2. logic to parse complex lists
3. support for complex constructors: `complex(tensor/int/float/bool, tensor/int/float/bool)`
4. Limited operator support
     - `add`, `sub`, `mul`, `torch.tensor`, `torch.as_tensor` 

**Follow-up work:**
1. Add complex support for unary and other registered ops.
2. support complex constructor with string as input (this is supported in Python eager mode).
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.
4. onnx doesn't support complex tensors, so we should error out with a clear and descriptive error message.

Differential Revision: [D27245059](https://our.internmc.facebook.com/intern/diff/D27245059)

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


**This PR adds:**
1. logic to parse complex constants (complex literals of the form `bj`)
2. logic to parse complex lists
3. support for complex constructors: `complex(tensor/int/float/bool, tensor/int/float/bool)`
4. Limited operator support
     - `add`, `sub`, `mul`, `torch.tensor`, `torch.as_tensor` 

**Follow-up work:**
1. Add complex support for unary and other registered ops.
2. support complex constructor with string as input (this is supported in Python eager mode).
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.
4. onnx doesn't support complex tensors, so we should error out with a clear and descriptive error message.

Differential Revision: [D27245059](https://our.internmc.facebook.com/intern/diff/D27245059)

[ghstack-poisoned]

SplitInfinity reviewed (approved):
I can't find any reason to not approve. Please make sure to fix failing tests before landing.

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


**This PR adds:**
1. logic to parse complex constants (complex literals of the form `bj`)
2. logic to parse complex lists
3. support for complex constructors: `complex(tensor/int/float/bool, tensor/int/float/bool)`
4. Limited operator support
     - `add`, `sub`, `mul`, `torch.tensor`, `torch.as_tensor` 

**Follow-up work:**
1. Add complex support for unary and other registered ops.
2. support complex constructor with string as input (this is supported in Python eager mode).
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.
4. onnx doesn't support complex tensors, so we should error out with a clear and descriptive error message.

Differential Revision: [D27245059](https://our.internmc.facebook.com/intern/diff/D27245059)

[ghstack-poisoned]

anjali411 committed:
Update on "Teach Python TS frontend to parse complex literals"


**This PR adds:**
1. logic to parse complex constants (complex literals of the form `bj`)
2. logic to parse complex lists
3. support for complex constructors: `complex(tensor/int/float/bool, tensor/int/float/bool)`
4. Limited operator support
     - `add`, `sub`, `mul`, `torch.tensor`, `torch.as_tensor` 

**Follow-up work:**
1. Add complex support for unary and other registered ops.
2. support complex constructor with string as input (this is supported in Python eager mode).
3. Test all emitXYZ for all XYZ in `ir_emitter.cpp` (currently only emitConst, emitValueToTensor are tested). e.g., test loops etc.
4. onnx doesn't support complex tensors, so we should error out with a clear and descriptive error message.

Differential Revision: [D27245059](https://our.internmc.facebook.com/intern/diff/D27245059)

[ghstack-poisoned]
